### DISCUSSION TITLE
Add Custom Attribute with Auto Instrumentation

### DISCUSSION THREAD

**Discussion by codeforaliving** [open-telemetry/opentelemetry-java-instrumentation#6458] [Q&A]
We are currently using Open Telemetry auto instrumentation with a Java extension (just a SpanProcessor for now) to rename span names as well as redact sensitive info from the url. Now we want to add a custom attribute to our traces and spans.  

We currently have a servlet filter to add the transaction id to the current span, but I am not sure how to propagate that value to child spans and I am not seeing the value at all in XRay. We are using Spring Boot and exporting through Prometheus. 

Any suggestions on how to get a custom attribute propagated?

**Comment by trask** [comment]
hi @codeforaliving! you could do something like https://github.com/microsoft/ApplicationInsights-Java/blob/main/agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/init/InheritedAttributesSpanProcessor.java

also check out https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/6191 and https://github.com/open-telemetry/oteps/pull/207

**Reply by codeforaliving** [reply]
@trask Thanks for the response.  We have a Java Spring Boot service that is using Open Telemetry auto-instrumentation.  The service project includes a common library that contains the extension with our custom SpanProcessor class (working well).  Each request into the service has a transaction id that is saved on the local thread.  How can I propagate that value on the parent and child spans within the context of the extension?

**Reply by trask** [reply]
in order to "communicate" between the application and the extension (which is in isolated class loader), you may need to put the ContextKey or ThreadLocal in the bootstrap class loader, or you maybe you could use Baggage which wouldn't require putting anything in the bootstrap class loader, see similar question here: https://github.com/open-telemetry/opentelemetry-java-instrumentation/discussions/5723#discussioncomment-2484980

**Reply by codeforaliving** [reply]
@trask Thanks for the reply.  I tried the same thing (using Context) as the user in the thread with the same result.  I am more of a Golang developer, so how do you put the ContextKey into the bootstrap class loader?

**Reply by codeforaliving** [reply]
@trask I had the same issue (using Context) as the user in the thread you referenced.  I ended up using Baggage to propagate a value, which is extracted and added as a span attribute in our custom span processor (extension).  However, now I am seeing increased latency with services that are utilizing Baggage.  Have you heard of issues with increase latency using Baggage?

**Reply by trask** [reply]
> Have you heard of issues with increase latency using Baggage?

I haven't, have you had any luck getting it sorted?

**Reply by codeforaliving** [reply]
Hello @trask. We are still working on a solution. Since we couldn’t pass a value through the context, we had to use baggage.

**Reply by codeforaliving** [reply]
Hello @trask .  Is this the correct way to add something to Baggage?

`Baggage.current().toBuilder().put("my-transaction-id", transactionId).build().storeInContext(Context.current()).makeCurrent();`

**Reply by trask** [reply]
make sure to use makeCurrent via try-with-resources to make sure that it is closed

**Reply by codeforaliving** [reply]
> make sure to use makeCurrent via try-with-resources to make sure that it is closed

@trask thanks for the response. So essentially wrap the entire call in a try-with-resources?  Will that still allow Baggage to be retrieved in our custom extension?

**Reply by codeforaliving** [reply]
@trask When I make the call via try-with-resources, I am unable to access the value from Baggage in my custom extension.  For example:

`try (Scope ignored = Context.current().with(Baggage.current().toBuilder().put("foo", "bar").build()).makeCurrent()) {
           // what should I do here?
  }`

**Reply by laurit** [reply]
> `try (Scope ignored = Context.current().with(Baggage.current().toBuilder().put("foo", "bar").build()).makeCurrent()) { // what should I do here? }`

Assuming you are using it in a servlet filter you should be calling `filterChain.doFilter(request, response);` there. This way the baggage you added in the context will be available during the request processing and will be cleared when the request is done.

**Comment by codeforaliving** [comment]
@laurit thanks for the response.  We have a class that adds a transaction ID to ThreadLocal and that is where I wanted to put the baggage logic. Can’t something just be added to baggage without the auto close?

**Reply by laurit** [reply]
We encourage people to use try with resource to ensure that the scope returned from `makeCurrent()` is closed. If you don't want to use try with resource then it is fine as long as you close the scope manually or verify yourself that you don't have a memory leak.

**Reply by codeforaliving** [reply]
@laurit Thanks again for the response.  Do you think it is sufficient to use the try-with-resources when extracting the Baggage items and adding them to a span in our custom SpanProcessor extension?  In other words, after we extract the Baggage from the parent context and determine that the value is there, perform the following:

```
...
final String transactionId = Baggage.fromContext(parentContext).getEntryValue("transaction-id");
try (Scope ignored = baggage.makeCurrent()) {
     span.setAttribute("transaction-id", transactionId);
}
...

```

**Reply by laurit** [reply]
I don't see how this could help. `Baggage.fromContext(parentContext)` finds the current baggage from context, no need to call  `makeCurrent` on it again. `span.setAttribute` doesn't do anything with context so changing the context with `makeCurrent` shouldn't be needed. Just
```
final String transactionId = Baggage.fromContext(parentContext).getEntryValue("transaction-id");
span.setAttribute("transaction-id", transactionId);
```
should be enough.
